### PR TITLE
Initial resource pack support

### PR DESCRIPTION
**Pull request by mathiascode**
The bare minimum for supporting server resource packs. Does not include any Lua APIs.

### CODE REVIEW

File: src/ClientHandle.h [hunk 1]
```diff
@@ -424,6 +428,11 @@ class cClientHandle  // tolua_export
 	AString m_Password;
 	Json::Value m_Properties;
 
+	/** The resource pack status sent from the client
+	Possible values are; 0: successfully loaded, 1: declined, 2: failed download, 3: accepted
+	Reference: https://wiki.vg/Protocol#Resource_Pack_Status */
+	int m_ResourcePackStatus = -1;
+
```
Comment (madmaxoft) [inline src/ClientHandle.h:435]: Make this an enum

File: src/Protocol/Protocol_1_9.cpp [hunk 2]
```diff
@@ -1233,6 +1233,28 @@ void cProtocol_1_9_0::SendResetTitle(void)
...
+	cPacketizer Pkt(*this, pktResourcePack);
+
+	cSha1Checksum Checksum;
+	Checksum.Update(reinterpret_cast<const Byte *>(a_ResourcePackUrl.c_str()), a_ResourcePackUrl.size());
+	Byte Digest[20];
+	Checksum.Finalize(Digest);
+	AString Sha1Output;
+	for (size_t i = 0; i < ARRAYCOUNT(Digest); i++)
+	{
+		AppendPrintf(Sha1Output, "", Digest[i]);
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_9.cpp:1247]: The format string seems erroneous here.

File: src/Protocol/Protocol_1_8.cpp [hunk 3]
```diff
@@ -1147,6 +1147,28 @@ void cProtocol_1_8_0::SendResetTitle(void)
...
+	cPacketizer Pkt(*this, pktResourcePack);
+
+	cSha1Checksum Checksum;
+	Checksum.Update(reinterpret_cast<const Byte *>(a_ResourcePackUrl.c_str()), a_ResourcePackUrl.size());
+	Byte Digest[20];
+	Checksum.Finalize(Digest);
+	AString Sha1Output;
+	for (size_t i = 0; i < ARRAYCOUNT(Digest); i++)
+	{
+		AppendPrintf(Sha1Output, "", Digest[i]);
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_8.cpp:1161]: The format string seems erroneous here.

File: src/Protocol/Protocol_1_9.cpp [hunk 4]
```diff
@@ -1233,6 +1233,28 @@ void cProtocol_1_9_0::SendResetTitle(void)
...
+	}
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_9.cpp:1248]: There's a `cSha1Checksum::DigestToJava` function, if that is not usable here, perhaps add a new one there, `DigestToHex`.

File: src/ClientHandle.cpp [hunk 5]
```diff
@@ -2228,7 +2235,10 @@ void cClientHandle::ServerTick(float a_Dt)
 
 	{
 		cCSLock lock(m_CSState);
-		if (m_State == csAuthenticated)
+		if (
+			(m_State == csAuthenticated) &&
+			(cRoot::Get()->GetServer()->GetResourcePackUrl().empty() || (m_ResourcePackStatus != -1))
+		)
```
Comment (madmaxoft) [inline src/ClientHandle.cpp:2241]: So when does the client start streaming chunks if resource pack url is not empty?

Comment (mathiascode) [reply src/ClientHandle.cpp:2241]: I'm starting to realize that this needs a different approach. Not exactly sure how to solve this, but as soon as chunks are streamed, the resource pack prompt closes.

**Comment by mathiascode** [comment]
What do you mean by multiple resource packs? What would the use case be? I don't think that's possible in the protocol.

File: src/ClientHandle.cpp (hunk 5 above)
Comment (mathiascode) [reply src/ClientHandle.cpp]: Fixed the issue.

File: src/mbedTLS++/Sha1Checksum.h [hunk 6]
```diff
@@ -32,6 +32,9 @@ class cSha1Checksum
 	/** Returns true if the object is accepts more input data, false if Finalize()-d (need to Restart()) */
 	bool DoesAcceptInput(void) const { return m_DoesAcceptInput; }
 
+	/** Converts a SHA1 digest into hex */
+	static void DigestToHex(const Checksum & a_Digest, AString & a_JavaOut);
```
Comment (bearbin) [inline src/mbedTLS++/Sha1Checksum.h:36]: `a_JavaOut`?

Comment (mathiascode) [reply src/mbedTLS++/Sha1Checksum.h:36]: Delicious copypasta.

**Comment by madmaxoft** [comment]
I have no idea what resourcepacks are. But the protocol basically allows to send multiple, so it's only a natural question to ask. If it makes no sense, then of course ignore me.

**Comment by madmaxoft** [comment]
And you should really install the CheckBasicStyle commit hook locally :)

File: src/mbedTLS++/Sha1Checksum.cpp [hunk 7]
```diff
@@ -85,6 +85,23 @@ void cSha1Checksum::Finalize(cSha1Checksum::Checksum & a_Output)
 
 
 
+void cSha1Checksum::DigestToHex(const Checksum & a_Digest, AString & a_Out)
+{
+	Checksum Digest;
+	memcpy(Digest, a_Digest, sizeof(Digest));
```
Comment (madmaxoft) [inline src/mbedTLS++/Sha1Checksum.cpp:91]: Why the copy? It should be possible to read out of `a_Digest` directly

**Comment by mathiascode** [comment]
Can that be automated somehow? I create too many local repos on a regular basis.

**Comment by madmaxoft** [comment]
I just have this in `$/.git/hooks/precommit`:
```
#!/bin/sh

# Redirect output to stderr.
exec 1>&2

exec lua.exe src/CheckBasicStyle.lua -g 
```

I assume you could just copy that file around to the new repositories. And why so many repos anyway?

**Comment by mathiascode** [comment]
Just working on a lot of things at once. I should spend some time on actually tidying up my dev environment though, it has become quite horrendous.

**Comment by tigerw** [comment]
> Just working on a lot of things at once

CPU joke lel

**Comment by mathiascode** [comment]
AMD gang.

**Comment by madmaxoft** [comment]
I think it'd be better to actually try compiling the stuff you're committing, that way you don't go from one thing to another again and again and again because the CI discovers simple errors in the code.

**Comment by mathiascode** [comment]
Well, it compiles under gcc. I should switch to clang.

**Comment by Bond-009** [comment]
> Is it possible to send multiple resource packs in the protocol? Should we allow that?

You can set multiple resource packs in the client, since a resource pack doesn't have to set textures for all blocks you can have it overriding only a couple of textures (e.g. redstone components or glass)

**Comment by Pokechu22** [comment]
While you can choose multiple packs clientside, I believe the server can only specify a single pack and it replaces the pack it's using if a new one is specified.  But I haven't actually tested that.

**Comment by mathiascode** [comment]
That's true, I forgot that the client allows you to combine multiple resource packs. From what I've read, it doesn't seem like you can add multiple resource packs to a server.

File: src/Protocol/Protocol_1_8.cpp [hunk 8]
```diff
@@ -1147,6 +1147,25 @@ void cProtocol_1_8_0::SendResetTitle(void)
 
 
 
+void cProtocol_1_8_0::SendResourcePack(const AString & a_ResourcePackUrl)
+{
+	cPacketizer Pkt(*this, pktResourcePack);
+
+	cSha1Checksum Checksum;
+	Checksum.Update(reinterpret_cast<const Byte *>(a_ResourcePackUrl.c_str()), a_ResourcePackUrl.size());
```
Comment (Pokechu22) [inline src/Protocol/Protocol_1_8.cpp:1155]: I'm pretty sure this is incorrect; it's supposed to be a checksum of the resource pack's contents (which vanilla has you specify in server.properties), not the URL.  I think the client uses it for caching, but I'm not 100% sure.

Comment (mathiascode) [reply src/Protocol/Protocol_1_8.cpp:1155]: Seems like you're right. At least it's not an urgent issue, but would be good for caching indeed.

**Pull request merged**